File: bindings/java/src/test/java/org/wickra/xray/GoldenTest.java
package org.wickra.xray;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;

// Cross-language golden parity: build the xray from each committed
// golden/specs/*.json, load the shared golden/data.json and read back the
// frame, then assert it equals golden/expected/<spec>.json byte-for-byte. The
// binding returns the core's compact command_json string verbatim, so byte
// equality is the exact cross-language parity check. The fixtures arrive in a
// later phase; until then the test is skipped.
class GoldenTest {
    private static Path findGolden() {
        Path dir = Path.of("").toAbsolutePath();
        for (int i = 0; i < 8 && dir != null; i++) {
            Path g = dir.resolve("golden");
            if (Files.isDirectory(g.resolve("specs"))) {
                return g;
            }
            dir = dir.getParent();
        }
        return null;
    }

    @Test
    void goldenFramesAreByteIdentical() throws IOException {
        Path golden = findGolden();
        assumeTrue(golden != null, "golden fixtures not present yet");

        String dataset = Files.readString(golden.resolve("data.json")).strip();
        String load = "{\"cmd\":\"load\",\"dataset\":" + dataset + "}";
        try (Stream<Path> specs = Files.list(golden.resolve("specs"))) {
            for (Path specPath : specs.filter(p -> p.toString().endsWith(".json")).toList()) {
                String spec = Files.readString(specPath);
                String name = specPath.getFileName().toString();
                String expected = Files.readString(golden.resolve("expected").resolve(name)).strip();
                try (Xray xray = new Xray(spec)) {
                    xray.command(load);
                    String raw = xray.command("{\"cmd\":\"frame\"}");
                    assertEquals(expected, raw.strip(), name);
                }
            }
        }
    }
}

File: bindings/node/__tests__/golden.test.js
"use strict";

// Cross-language golden parity: build the xray from each committed
// `golden/specs/*.json`, load the shared `golden/data.json` and read back the
// `frame`, then assert it equals `golden/expected/<spec>.json` byte-for-byte.
// Because every binding returns the core's compact `command_json` string
// verbatim, byte equality is the exact cross-language parity check. The fixtures
// arrive in a later phase; until then this test skips cleanly.

const { test } = require("node:test");
const assert = require("node:assert");
const fs = require("node:fs");
const path = require("node:path");
const { Xray } = require("../index.js");

function findGolden() {
  let dir = __dirname;
  for (let i = 0; i < 8; i++) {
    const g = path.join(dir, "golden");
    if (fs.existsSync(path.join(g, "specs"))) {
      return g;
    }
    dir = path.dirname(dir);
  }
  return null;
}

test("golden frames are byte-identical", (t) => {
  const golden = findGolden();
  if (!golden) {
    t.skip("golden fixtures not present yet");
    return;
  }
  const dataset = fs.readFileSync(path.join(golden, "data.json"), "utf8");
  const specDir = path.join(golden, "specs");
  for (const file of fs.readdirSync(specDir).filter((f) => f.endsWith(".json"))) {
    const spec = fs.readFileSync(path.join(specDir, file), "utf8");
    const expected = fs
      .readFileSync(path.join(golden, "expected", file), "utf8")
      .trim();
    const xray = new Xray(spec);
    xray.command(JSON.stringify({ cmd: "load", dataset: JSON.parse(dataset) }));
    const response = xray.command(JSON.stringify({ cmd: "frame" }));
    assert.strictEqual(response.trim(), expected, `mismatch for ${file}`);
  }
});
